Adds Error summary at the end of resource section

ashish111333 opened the pull request:
Adds a line at the end of  resources showing resources that errored.
closes #11972

ashish111333 commented:
/run-acceptance-tests

tgummerer commented:
This should have a changelog entry. Also a screenshot of before and after would be helpful for reviewing the change here.

ashish111333 commented:
so i have to update this change in changelog folder ?

tgummerer commented:
You can run `make changelog`, and it will guide you through the process.  Also read https://github.com/pulumi/pulumi/blob/master/CONTRIBUTING.md#changelog-messages

tgummerer commented on changelog/pending/20250922--cli-display--report-error-count-and-deploymnt-message-in-diag-sec.yaml:
```diff
@@ -0,0 +1,4 @@
+changes:
+- type: feat
+  scope: cli/display
+  description: adds a line at end of diag sec showing whether dep was success or not with error count
```
```suggestion
  description: Add a line at end of diag section showing whether deployment was success or not with error count
```

Please follow the guidelines in the document I linked.  Also this changelog is user facing, so it shouldn't use abbreviations, especially ones that can be confused with others, such as `dep`, which could mean dependency or deployment.

ashish111333 commented on pkg/backend/display/progress.go:
```diff
@@ -816,6 +821,16 @@ func (display *ProgressDisplay) printDiagnostics() {
+	// deployment was successful or not
+	if errDiagMessgCount > 0 {
+		errCountStr := fmt.Sprintf("[error count = %d]", errDiagMessgCount)
+		display.println(colors.BrightRed + "  " + "deployment encountered errors " + errCountStr + colors.Reset)
+	} else {
+		display.println(colors.BrightGreen + "  " + "deployment completed" + colors.Reset)
```
I added this because  that means the latest change you made your infra code  is done without  any errors so I am using the world deployment in that sense.

tgummerer commented on changelog/pending/20250924--cli-display--added-error-summary-in-resource-summary.yaml:
```diff
@@ -0,0 +1,4 @@
+changes:
+- type: feat
+  scope: cli/display
+  description: Adds resources error summary at the end of resource section
```
```suggestion
  description: Add resources error summary at the end of resource section
```

tgummerer commented on pkg/backend/display/progress.go:
```diff
@@ -1027,6 +1027,28 @@ func (display *ProgressDisplay) printSummary() {
+
+	// after printing resources changes summary also add the count for resources that errored
+
+	// track Resources that errored
+	resourcesErrored := 0
+
+	// get resource rows
+	rr := toResourceRows(display.eventUrnToResourceRow, display.opts.DeterministicOutput)
+
+	// check error count for each resource
+	for _, r := range rr {
+		if r.DiagInfo().ErrorCount > 0 {
+			resourcesErrored++
+		}
+	}
+
+	// finally print the line showing resources errored
```
These comments are superflous.  The code tells us what's going on here.  A good comment describes *why* something happens, not *what* happens in the code.  We can just get rid of them here.

ashish111333 replied:
I did  think  that I am writing too much. ok will remove them.

tgummerer reviewed (commented):
Pretty sure this is going to fail some display tests that need to be fixed up.  You can do that by running the tests locally with `PULUMI_ACCEPT=true` set in the environment.

ashish111333 commented:
@tgummerer  here's a ss  of  how things will look. The error summary  will always be shown at bottom under Resources (img below ) --

<img width="1067" height="535" alt="pul_err2" src="https://github.com/user-attachments/assets/e4ca885a-461c-4745-8e22-792a6264e93a" />

tgummerer commented:
/run-acceptance-tests

ashish111333 commented on pkg/backend/display/testdata/not-truncated/failed-stacks-dont-rewrite-operations.json.interactive-100x80-cooked.stdout.txt:
```diff
@@ -10,6 +10,5 @@
 
 <{%fg 13%}><{%bold%}>Resources:<{%reset%}>
     1 unchanged
-
```
this new line causes a gap between error summary and summary changes, it's fine I think.

ashish111333 replied:
this new line causes a gap between error summary and summary changes, it's fine I think.

tgummerer replied:
Well no, this is the existing output changing, when it really shouldn't.  There's no errors here, so there shouldn't be any changes in the output.

ashish111333 replied:
@tgummerer I removed a new line from the  function ( renderSummary )  that runs before my code it prints resources changes now this func adds a new line I just removed this new line and added it at the end my code. It's working fine as expected. I didn't want to pollute the renderSummary function so it's totally fine.

ashish111333 replied:
@tgummerer  the change is totally safe and doing what was discussed throughout.

tgummerer replied:
Again no.  The output here *should not change*. If there are no errors, with your change we're missing a newline that should be there.  The change is not "safe", it worsens the output we're getting by removing that newline if there are no errors, and can not be merged in the current state.

ashish111333 commented on pkg/backend/display/diff.go:
```diff
@@ -283,7 +283,6 @@ func renderSummaryEvent(event engine.SummaryEventPayload, diffStyleSummary bool,
 			out.WriteString(opts.Color.Colorize(piece))
 		}
 
-		fprintIgnoreError(out, "\n")
```
this is how it will look like if i add that line again 

<img width="857" height="483" alt="pul_err3" src="https://github.com/user-attachments/assets/685a0e0b-3bba-4f6d-9b94-e50172339055" />

see the gap b/w resource changes and error summary

tgummerer commented on pkg/backend/display/progress.go:
```diff
@@ -1027,6 +1027,23 @@ func (display *ProgressDisplay) printSummary() {
+	if resourcesErrored > 0 {
+		errSummaryStr := fmt.Sprintf("%d errored", resourcesErrored)
+		display.println("    " + colors.Red + errSummaryStr + colors.Reset)
+		display.println("")
```
I think this needs to be moved out of the `if`, to replace the line you removed here: https://github.com/pulumi/pulumi/pull/20534/files#diff-4d8b36285a68d688efdec3d43b7d4f77fb0f345e8075fb509fd6a5a56d66626e

tgummerer commented:
Btw, I think your PR also needs rebasing onto the latest `master`, there's unfortunately a few conflicts here.

ashish111333 commented:
can you run Tests now

tgummerer commented on pkg/backend/display/testdata/not-truncated/failed-stacks-dont-rewrite-operations.json.interactive-100x80-cooked.stdout.txt:
```diff
@@ -1,11 +1,15 @@
-
 <{%fg 13%}><{%bold%}>Duration:<{%reset%}> 8s
 
+
```
Huh, we have an additional line here, but the line above the "Duration:" header is still removed.  This is still wrong.  Again, these files should not be changing at all.

ashish111333 replied:
I am about to cry now :sob:

tgummerer commented:
/run-acceptance-tests

ashish111333 commented:
@tgummerer  the last commit should fix. please run Tests.

tgummerer commented:
/run-acceptance-tests

ashish111333 commented on pkg/engine/lifecycletest/testdata/output/TestBadResourceType/-0-0/progress.stdout.txt:
```diff
@@ -13,6 +13,6 @@
     <{%fg 2%}>+ 2 created<{%reset%}>
-
+    <{%fg 1%}>1 errored<{%reset%}>
 <{%fg 13%}><{%bold%}>Duration:<{%reset%}> 1s
```
fixed.

tgummerer commented:
/run-acceptance-tests

ashish111333 commented on pkg/backend/display/diff.go:
```diff
@@ -81,9 +83,16 @@ func ShowDiffEvents(op string, events <-chan engine.Event, done chan<- bool, opt
 				if payload.Severity == diag.Error || payload.Severity == diag.Warning {
 					out = stderr
 				}
+				if payload.Severity == diag.Error {
+					resourcesErrored[payload.URN] = true
+				}
```
That's why I mapped to these diag errors to single resource URNs is that in correct?

ashish111333 commented on pkg/engine/lifecycletest/testdata/output/TestAliasURNs/urn-fail-duplicate-0-0-preview/progress.stdout.txt:
```diff
@@ -13,4 +13,5 @@
 <{%fg 13%}><{%bold%}>Resources:<{%reset%}>
     <{%fg 3%}>~ 1 to update<{%reset%}>
-
+    <{%fg 1%}>1 errored
```
I will move new line out of colors would put it at end of  error summary after color reset.

ashish111333 commented:
@tgummerer display Tests have failed (expected ) but  is the output fine now  you according you ?

tgummerer commented:
How is that expected?  If they fail the output is not fine.

ashish111333 commented:
@tgummerer  I  reverted all commits in my local branch pushed the the last change as a single commit this commit contains all golden files after fixing engine and display tests  ( now the display testdata is showing in commit that  one golden file was changed after I fixed tests) . Only one display test failed   I fixed it and the updated golden file is in the commit this should work.

tgummerer commented:
/run-acceptance-tests

tgummerer commented on pkg/backend/display/testdata/not-truncated/failed-stacks-dont-rewrite-operations.json.stdout.txt:
```diff
@@ -9,5 +9,7 @@
 <{%reset%}><{%reset%}><{%fg 13%}><{%bold%}>Resources:<{%reset%}>
     1 unchanged
+    <{%fg 1%}>1 errored<{%reset%}>
+
```
There's a newline too many here if there are errored resources.  It should have the same amount of newlines (1) no matter if there is an errored resource or if there is not.

tgummerer commented:
/run-acceptance-tests

ashish111333 commented:
@tgummerer  Run test please.

tgummerer commented:
/run-acceptance-tests
